share/shwap/p2p/bitswap/bitswap.go

```diff
@@ -0,0 +1,72 @@
+package bitswap
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	logger "github.com/ipfs/go-log/v2"
+
+	"github.com/celestiaorg/celestia-node/share"
+)
+
+// TODO
+//  * Try generic hasher
+//  * Godoc
+
+var log = logger.Logger("shwap/bitswap")
+
+var globalRootRegistry rootIndex
+
+func addRoot(height uint64, root *share.Root) {
+	globalRootRegistry.Add(height, root)
+}
+
+func getRoot(height uint64) (*share.Root, error) {
+	return globalRootRegistry.Get(height)
+}
+
+func deleteRoot(height uint64) {
+	globalRootRegistry.Delete(height)
+}
+
+type rootIndex struct {
+	mp sync.Map
+}
+
+type rootValue struct {
+	*share.Root
+	refs atomic.Int32
+}
+
+func (roots *rootIndex) Add(height uint64, root *share.Root) {
+	val, _ := roots.mp.LoadOrStore(height, &rootValue{Root: root})
+	rootVal := val.(*rootValue)
+	rootVal.refs.Add(1)
+}
+
+func (roots *rootIndex) Get(height uint64) (*share.Root, error) {
+	val, ok := roots.mp.Load(height)
+	if !ok {
+		return nil, fmt.Errorf("no Root registered")
+	}
+	rootVal := val.(*rootValue)
+
+	return rootVal.Root, nil
+}
+
+func (roots *rootIndex) Delete(height uint64) {
+	val, ok := roots.mp.Load(height)
+	if !ok {
+		return
+	}
+	rootVal := val.(*rootValue)
+
+	refs := rootVal.refs.Add(-1)
+	switch {
+	case refs == 0:
+		roots.mp.Delete(height)
+	case refs < 0:
+		panic("negative root reference count")
+	}
```

Comment:
Its ok in this case to use just Delete. The item either exists or not. The first one will Delete it and subsequejt one will be noop

Reply:
The deletion was atomic in this case, because the value is always the same and either exist or not. If Delete executed over the deleted values its noop

Reply:
But doesn't matter much, because the design has changed